### PR TITLE
feat(okf): migrate AIMS knowledge bundle to v0.2

### LINKED ISSUE #160: Migrate the AIMS OKF bundle and tooling to OKF v0.2
## Context

The canonical AIMS knowledge bundle currently declares `okf_version: "0.1"`, and the validator, authoring skills, curator output, and documentation are built around the OKF v0.1 conventions.

The upstream Open Knowledge Format specification has advanced to **v0.2**:

- https://github.com/GoogleCloudPlatform/knowledge-catalog/blob/main/okf/SPEC.md

The migration has two breaking changes:

- replace legacy `timestamp` with `generated: { by, at }`
- replace body-level `# Citations` lists with frontmatter `sources`

OKF v0.2 also standardizes provenance, verification, lifecycle, freshness, and attested-computation metadata.

## Scope

- Update `okf/index.md` to declare OKF v0.2.
- Migrate every concept in `okf/` to v0.2-compatible frontmatter:
  - replace `timestamp` with `generated.by` and `generated.at`
  - move repository/document provenance from `# Citations` into `sources`
  - normalize `status` to the v0.2 lifecycle values: `draft`, `stable`, or `deprecated`
  - review the current structured `resource` values and make them compatible with the v0.2 `resource` semantics
  - retain AIMS-specific metadata only as valid extension fields
- Update `src/aims/okf_hugo.py` and `tools/okf_hugo_adapter.py` for v0.2 validation and Hugo rendering.
- Update the curator-generated concept skeletons in `src/aims/okf_curation.py` to emit v0.2 metadata.
- Preserve v0.2 provenance, trust, lifecycle, and extension fields in generated Hugo metadata without flattening or dropping them.
- Update OKF authoring, curation, publication, and PR-review skills plus repository documentation to reference v0.2 rather than the v0.1 draft.
- Regenerate `content/knowledge/` and update unit tests, fixtures, and golden files.

## Acceptance criteria

- `okf/index.md` declares `okf_version: "0.2"`.
- No canonical concept uses legacy `timestamp` or a body-level `# Citations` section.
- All lifecycle statuses conform to `draft | stable | deprecated`.
- New curator proposals emit valid v0.2 concepts.
- The adapter detects invalid v0.2 metadata and renders all supported metadata deterministically.
- Internal OKF links and generated Hugo URLs remain unchanged unless a migration requires an intentional documented change.
- The following checks pass:

```console
uv run python tools/okf_hugo_adapter.py --src okf --dst content/knowledge --check
uv run pytest
hugo --gc --minify
```

## Non-goals

- Introducing an external schema registry or bespoke OKF runtime.
- Converting quantitative analysis outputs into OKF source-of-truth documents.
- Adding attested computations unless an existing AIMS concept has a concrete deterministic computation that benefits from the v0.2 model.

### PR DESCRIPTION
**Pull request by dceoy**
## Summary

- migrate the canonical OKF bundle from v0.1 to v0.2 `generated`, `sources`, and lifecycle metadata
- validate and deterministically preserve v0.2 provenance, trust, freshness, computation, and extension fields in Hugo shadow content
- update curator proposals, repository skills and docs, tests and fixtures, and regenerate `content/knowledge/`

## Validation

- `.agents/skills/local-qa/scripts/qa.sh`
- 1,142 tests with 100% branch coverage
- `uv run python tools/okf_hugo_adapter.py --src okf --dst content/knowledge --check`
- `hugo --gc --minify`
- security pattern scan and parser/rendering review: no actionable findings

Closes #160

### CODE REVIEW

File: src/aims/okf_hugo.py [hunk 1]
```diff
@@ -32,10 +33,14 @@ class OkfYamlLoader(yaml.SafeLoader):
 FRONT_MATTER = re.compile(r"\A---\n(?P<meta>.*?)\n---\n(?P<body>.*)\Z", re.DOTALL)
 LINK = re.compile(r"(?<!!)\[(?P<label>[^\]]+)\]\((?P<target>[^)]+)\)")
 TAG = re.compile(r"^[a-z0-9]+(?:-[a-z0-9]+)*$")
-RECOMMENDED_CONCEPT_FIELDS = ("title", "description", "tags", "timestamp")
+ACTOR = re.compile(r"^(?:[a-z][a-z0-9-]*:[^\s:]+|[^/\s]+/[^/\s]+)$")
+CITATIONS_HEADING = re.compile(r"(?m)^# Citations\s*$")
+ISO_DATE = re.compile(r"^\d{4}-\d{2}-\d{2}$")
+RECOMMENDED_CONCEPT_FIELDS = ("title", "description", "tags", "generated")
```
[the remark this answers is not in the document]
Comment (dceoy) [reply src/aims/okf_hugo.py]: Added `sources` to `RECOMMENDED_CONCEPT_FIELDS` (src/aims/okf_hugo.py:39) so `--check` now flags concepts missing it; `validate_v02_metadata` is left untouched since it also runs on reserved/non-concept docs where sources isn't applicable.

[the remark this answers is not in the document]
Comment (dceoy) [reply src/aims/okf_hugo.py]: Also require the list to be non-empty. Adding `sources` to `RECOMMENDED_CONCEPT_FIELDS` would reject an omitted key, but `sources: []` still passes `_validate_sources()` because it only checks that the value is a list. Since AIMS is making provenance mandatory, validate both presence and at least one source entry.

**Review by dceoy — COMMENTED** [review]
The v0.2 migration is generally well covered, but the provenance references are not portable OKF paths and the existing `sources` validation gap remains unresolved. These should be addressed before merging.

File: okf/concepts/architecture.md [hunk 2]
```diff
@@ -4,11 +4,17 @@ title: AIMS Architecture
...
-status: seeded
+generated:
+  by: process:aims-okf-migration
+  at: 2026-07-27T00:00:00Z
+status: stable
+sources:
+  - id: repository-guidelines
+    resource: AGENTS.md
+    title: Repository guidelines
+  - id: repository-readme
```
Comment (dceoy) [inline okf/concepts/architecture.md:15]: **[P2] Use resolvable resources for repository-level sources**

`AGENTS.md` and the other migrated values such as `README.md`, `OPERATIONS.md`, `.github/workflows/...`, and `data/...` are repository-root paths, not resources inside the OKF bundle. An OKF consumer resolving this relative value from `okf/concepts/architecture.md` cannot reach the intended file (and the file is outside the distributed `okf/` bundle in any case), so the new provenance metadata is not portable or followable. Use an absolute repository URL/URI, mirror the material under an OKF `references/` concept, or otherwise emit a resource that remains resolvable when the bundle is consumed independently. Update the curator skeleton too, since it currently emits the same repository-root-relative form for qualitative artifacts.

Comment (dceoy) [reply okf/concepts/architecture.md:15]: Switched `resource` fields to absolute GitHub blob URLs (e.g. `https://github.com/dceoy/aims/blob/main/AGENTS.md`) across all migrated concepts, and updated `concept_skeleton()` in `src/aims/okf_curation.py` to emit the same URL form for qualitative-artifact sources (commit 551c601).

File: src/aims/okf_hugo.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (dceoy) [reply src/aims/okf_hugo.py]: Closed the remaining gap: `_validate_sources` now rejects an empty `sources` list, not just a missing key (commit 551c601).

File: src/aims/okf_hugo.py [hunk 3]
```diff
@@ -336,6 +341,262 @@ def validate_reserved(document: OkfDocument, src_root: Path) -> list[str]:
...
+        )
+    if "attester" in metadata:
+        errors.extend(
+            _validate_resource_mapping(
+                document, "attester", metadata["attester"], receipt=False
+            )
+        )
+    if metadata.get("type") == "Attested Computation" and not _is_non_empty_string(
+        metadata.get("runtime")
+    ):
```
Comment (dceoy) [inline src/aims/okf_hugo.py]: **[P2] Require an executable computation contract**

This branch only requires `runtime`, so an `Attested Computation` with neither a `computation` path nor a body `# Computation` fence passes `--check` even though there is nothing to execute. OKF v0.2 defines the computation as either the referenced file or the body fence. Validate that exactly one usable form is present (or at minimum that one is present), and add rejection tests for an empty contract.

File: src/aims/okf_hugo.py [hunk 4]
```diff
@@ -336,6 +341,262 @@ def validate_reserved(document: OkfDocument, src_root: Path) -> list[str]:
...
+        ):
+            errors.append(
+                _metadata_error(
+                    document,
+                    f"'{field}.author' must follow the OKF v0.2 actor convention",
+                )
+            )
+        if "usage_count" in source and (
+            not isinstance(source["usage_count"], int)
+            or isinstance(source["usage_count"], bool)
```
Comment (dceoy) [inline src/aims/okf_hugo.py]: **[P2] Require a usage window for usage counts**

The validator accepts `usage_count` after checking only its type and range. Without either this source's `usage_window` or the document-level shared `usage_window`, the count has no defined observation period and cannot be interpreted as the credibility signal specified by OKF v0.2. When a source has `usage_count`, require one of those windows; also validate `from <= to` so reversed periods do not pass.

**Review by dceoy — COMMENTED** [review]
The v0.2 migration is well covered and the previous provenance findings are resolved. Two validation gaps remain: Attested Computation concepts can still omit the computation itself, and usage counts can be accepted without the window that gives them meaning. Please address these before merging.

File: src/aims/okf_hugo.py (hunk 3 above)
Comment (dceoy) [reply src/aims/okf_hugo.py]: Fixed: an Attested Computation now requires exactly one of a non-empty `computation` path or a non-empty body `# Computation` fenced code block (rejects missing, empty, and both-present cases), with tests added (commit b1ee897).

File: src/aims/okf_hugo.py (hunk 4 above)
Comment (dceoy) [reply src/aims/okf_hugo.py]: Fixed: sources with `usage_count` now require a source-level or document-level `usage_window`, and `usage_window` now rejects `from > to`, with tests added (commit b1ee897).

File: src/aims/okf_hugo.py [hunk 5]
```diff
@@ -336,6 +344,299 @@ def validate_reserved(document: OkfDocument, src_root: Path) -> list[str]:
...
+        errors.append(
+            _metadata_error(
+                document, "body-level '# Citations' is superseded by 'sources'"
+            )
+        )
+    if "resource" in metadata and not _is_non_empty_string(metadata["resource"]):
+        errors.append(
+            _metadata_error(document, "'resource' must be a non-empty URI or path")
+        )
+    if "status" in metadata and metadata["status"] not in LIFECYCLE_STATUSES:
```
Comment (dceoy) [inline src/aims/okf_hugo.py]: **[P2] Reject non-string status values without crashing**

Membership against the set raises `TypeError: unhashable type` when YAML supplies a list or mapping, for example `status: [stable]`. That makes `--check` terminate with a traceback instead of returning a path-specific validation error. Check that the value is a string before testing membership, and add list/mapping regression cases.

File: src/aims/okf_hugo.py [hunk 6]
```diff
@@ -336,6 +344,299 @@ def validate_reserved(document: OkfDocument, src_root: Path) -> list[str]:
...
+    if metadata.get("type") == "Attested Computation":
+        if not _is_non_empty_string(metadata.get("runtime")):
+            errors.append(
+                _metadata_error(
+                    document,
+                    "'runtime' is required for an Attested Computation",
+                )
+            )
+        has_computation_file = _is_non_empty_string(metadata.get("computation"))
+        has_computation_body = _has_computation_body(document.body)
```
Comment (dceoy) [inline src/aims/okf_hugo.py]: **[P2] Reject multiple inline computation blocks**

`_has_computation_body()` reduces the first regex match to a boolean, so a `# Computation` section containing two non-empty fenced blocks passes validation. OKF v0.2 defines the inline form as a single fenced code block; accepting multiple blocks leaves the sanctioned computation ambiguous to executors and attesters. Count matching computation blocks and require exactly one when the inline form is used.

**Review by dceoy — COMMENTED** [review]
The OKF v0.2 migration is broadly well covered and the latest CI run passes, but the validator still has three boundary-input gaps that should be addressed before merging.

**[P2] Require `type` to be a string**

`validate_reserved_or_concept()` currently converts the value with `str(...)`, so values such as `type: [concept]` or `type: {name: concept}` pass the non-empty check and are preserved into `params.okf_type`. OKF v0.2 defines `type` as a short string; validate `isinstance(value, str)` and non-emptiness, and add list/mapping regression tests.

File: src/aims/okf_hugo.py (hunk 5 above)
Comment (dceoy) [reply src/aims/okf_hugo.py]: Guarded the membership check with an explicit `isinstance(..., str)` check so list/mapping `status` values now produce a validation error instead of a `TypeError`; added list and mapping regression cases (commit 44c763f).

File: src/aims/okf_hugo.py (hunk 6 above)
Comment (dceoy) [reply src/aims/okf_hugo.py]: Now count consecutive fenced blocks after the `# Computation` heading and reject when more than one non-empty block is present as ambiguous (commit 44c763f).

File: src/aims/okf_hugo.py [hunk 7]
```diff
@@ -313,11 +318,13 @@ def validate_reserved_or_concept(document: OkfDocument, src_root: Path) -> list[
             "OKF conformance error: "
```
Comment (dceoy) [inline src/aims/okf_hugo.py:321]: **[P2] Reject non-string `type` values**

This still coerces `type` with `str(...)`, so YAML values such as `type: [concept]` or `type: {name: concept}` pass the non-empty check and are emitted as structured `params.okf_type`. OKF v0.2 requires `type` to be a short string. Validate `isinstance(value, str)` before checking non-emptiness, and add list/mapping regression cases.

File: src/aims/okf_hugo.py [hunk 8]
```diff
@@ -336,6 +343,321 @@ def validate_reserved(document: OkfDocument, src_root: Path) -> list[str]:
...
+    heading = COMPUTATION_HEADING.search(body)
+    if not heading:
+        return 0
+    position = heading.end()
+    count = 0
+    while True:
+        match = CONSECUTIVE_FENCED_BLOCK.match(body, position)
+        if not match:
+            return count
+        if match.group("code").strip():
```
Comment (dceoy) [inline src/aims/okf_hugo.py]: **[P2] Count all fences in the computation section**

The new loop stops as soon as the next token is not another fence, so a second non-empty fenced block separated by prose still passes as a single computation, for example `fence -> explanation -> fence`. The inline contract is defined as one fenced code block under `# Computation`; scan until the next H1/end of body and reject any second fence, with a regression test that inserts prose between the two blocks.

**Review by dceoy — COMMENTED** [review]
The migration and latest CI run are otherwise in good shape, but two validator gaps remain on the current head (`44c763f`) and should be addressed before merging.

**Review by dceoy — COMMENTED** [review]
The v0.2 migration is broadly consistent with the upstream specification and the latest CI run succeeds. Two remaining boundary cases can produce broken provenance or incomplete computation validation and should be addressed before merge.

File: src/aims/okf_curation.py [hunk 9]
```diff
@@ -281,7 +282,14 @@ def concept_skeleton(cluster: ThemeCluster, as_of: str) -> str:
...
     ]
     sources = sorted({occurrence.source for occurrence in cluster.occurrences})
-    citations = [f"- `{source}`" for source in sources]
+    source_metadata = [
+        line
+        for source in sources
+        for line in (
+            f"  - resource: {REPO_BLOB_URL}/{source}",
+            f"    title: Qualitative analysis artifact {Path(source).stem}",
+        )
```
Comment (dceoy) [inline src/aims/okf_curation.py:292]: **[P2] Keep curator source URLs repository-relative**

`load_artifacts()` stores `path.as_posix()`, and the CLI explicitly accepts absolute `--qualitative-dir` paths (the CLI test uses one). In that case `source` becomes something like `/tmp/.../2024-01-01.json`, so this interpolation emits `https://github.com/dceoy/aims/blob/main//tmp/...`, not the artifact in the repository. Since these new `sources` entries replace the body citations and are the concept's provenance, normalize artifact paths relative to the repository root (or carry a separate canonical repository path) before constructing the URL, and assert the generated resource in the CLI test.

File: src/aims/okf_hugo.py [hunk 10]
```diff
@@ -336,6 +342,329 @@ def validate_reserved(document: OkfDocument, src_root: Path) -> list[str]:
...
+    for line in body[heading.end() :].splitlines():
+        if in_fence:
+            if line.startswith("```"):
+                if "".join(code_lines).strip():
+                    count += 1
+                in_fence = False
+                code_lines = []
+            else:
+                code_lines.append(line)
+        elif line.startswith("```"):
```
Comment (dceoy) [inline src/aims/okf_hugo.py]: **[P2] Validate all computation sections with Markdown fence rules**

The scanner starts at the first `# Computation` and stops at the next H1, so `# Computation` + fence, then `# Notes`, then a second `# Computation` + fence still passes with a count of one. It also recognizes only column-zero backtick fences, rejecting standard Markdown tilde fences or fences indented by up to three spaces. OKF defines the body as standard Markdown and the inline contract as a single fenced code block. Scan every `# Computation` section using Markdown fence semantics, and add regressions for duplicate sections plus tilde/indented fences.

File: src/aims/okf_curation.py (hunk 9 above)
Comment (dceoy) [reply src/aims/okf_curation.py:292]: Normalized citation to the canonical repository path: `load_artifacts()` now cites `data/qualitative/<file>` regardless of the `--qualitative-dir` argument's form, so an absolute dir can no longer leak into the generated resource URL; added a regression asserting the emitted resource and that the tmp dir path is absent (commit 58bd173).

File: src/aims/okf_hugo.py (hunk 10 above)
Comment (dceoy) [reply src/aims/okf_hugo.py]: Fixed: the scanner now walks every '# Computation' heading (not just the first) and recognizes tilde fences plus fences indented up to 3 spaces per CommonMark; fence state is tracked independent of heading detection, so a fenced example containing a literal '# Computation' or '# Notes' line is never mistaken for a real heading. Added regressions for duplicate sections, tilde fences, indented fences, and a heading-like line inside an unrelated fence (commit 58bd173).

File: src/aims/okf_hugo.py [hunk 11]
```diff
@@ -32,10 +33,17 @@ class OkfYamlLoader(yaml.SafeLoader):
 FRONT_MATTER = re.compile(r"\A---\n(?P<meta>.*?)\n---\n(?P<body>.*)\Z", re.DOTALL)
 LINK = re.compile(r"(?<!!)\[(?P<label>[^\]]+)\]\((?P<target>[^)]+)\)")
 TAG = re.compile(r"^[a-z0-9]+(?:-[a-z0-9]+)*$")
-RECOMMENDED_CONCEPT_FIELDS = ("title", "description", "tags", "timestamp")
+ACTOR = re.compile(r"^(?:[a-z][a-z0-9-]*:[^\s:]+|[^/\s]+/[^/\s]+)$")
+CITATIONS_HEADING = re.compile(r"(?m)^# Citations\s*$")
+COMPUTATION_HEADING_LINE = re.compile(r"^# Computation\s*$")
+HEADING_LINE = re.compile(r"^#[ \t]")
+FENCE_LINE = re.compile(r"^ {0,3}(?P<marker>`{3,}|~{3,})")
+ISO_DATE = re.compile(r"^\d{4}-\d{2}-\d{2}$")
```
Comment (dceoy) [inline src/aims/okf_hugo.py]: **[P2] Parse Markdown H1 headings consistently**

These three regexes only recognize column-zero ATX headings, and the citations check scans raw fenced content. Standard Markdown also recognizes up to three leading spaces and an optional closing `#` sequence. Consequently, `   # Computation` is reported as missing, an indented next H1 does not end the computation section, `# Citations #` bypasses the legacy-section guard, and a literal `# Citations` inside a fenced example is rejected. Use one fence-aware H1 scanner for computation boundaries and citation detection, and add regression cases for these forms.

File: src/aims/okf_hugo.py [hunk 12]
```diff
@@ -32,10 +33,17 @@ class OkfYamlLoader(yaml.SafeLoader):
...
+RECOMMENDED_CONCEPT_FIELDS = ("title", "description", "tags", "generated", "sources")
```
Comment (dceoy) [inline src/aims/okf_hugo.py:44]: **[P2] Validate closing-fence syntax separately**

`FENCE_LINE` accepts arbitrary text after the marker and is reused to detect closing fences. Markdown closing fences may only have trailing whitespace, so a line such as `````not-a-close`` currently terminates the block. That can make an unterminated computation appear valid and expose following heading-like lines to the section scanner. Use a stricter closing-fence check (same marker and length, whitespace-only remainder) and add a malformed-closer regression test.

**Review by dceoy — COMMENTED** [review]
The migration is otherwise consistent and CI passes, but two Markdown-parser boundary cases remain in the new v0.2 validator. They can either reject valid concepts or let malformed/legacy sections pass, so they should be addressed before merge.

File: src/aims/okf_hugo.py (hunk 11 above)
Comment (dceoy) [reply src/aims/okf_hugo.py]: Fixed: computation-boundary and citation-heading detection now share one fence-aware ATX H1 scanner (0-3 space indent, optional closing `#` sequence) in `_iter_body_lines`, with regression cases for indented headings and citations-in-fences (commit 4163757).

File: src/aims/okf_hugo.py (hunk 12 above)
Comment (dceoy) [reply src/aims/okf_hugo.py:44]: Fixed: closing fences now require the same marker char, length >= opening, and a whitespace-only remainder, via a shared `_closing_fence` helper used by both the section scanner and `_fenced_code_blocks`; a malformed closer now correctly leaves the computation unterminated (commit 4163757).

File: src/aims/okf_hugo.py [hunk 13]
```diff
@@ -336,6 +346,413 @@ def validate_reserved(document: OkfDocument, src_root: Path) -> list[str]:
...
+                or isinstance(source["usage_count"], bool)
+                or source["usage_count"] < 0
+            ):
+                errors.append(
+                    _metadata_error(
+                        document,
+                        f"'{field}.usage_count' must be a non-negative integer",
+                    )
+                )
+            if "usage_window" not in source and "usage_window" not in document.metadata:
```
Comment (dceoy) [inline src/aims/okf_hugo.py]: **[P2] Require the document-level usage window defined by OKF v0.2**

OKF v0.2 defines `usage_window` as a sibling of `sources`; it is not a field on an individual source entry. This condition lets `sources[].usage_window` satisfy `usage_count`, so a document with no top-level window passes `--check`, but a conforming consumer will read the count without its observation period. Require `document.metadata["usage_window"]` whenever a source has `usage_count`; a nested extension may be preserved, but should not substitute for the standardized field. See SPEC.md §5.1.

File: src/aims/okf_hugo.py [hunk 14]
```diff
@@ -32,10 +36,16 @@ class OkfYamlLoader(yaml.SafeLoader):
 FRONT_MATTER = re.compile(r"\A---\n(?P<meta>.*?)\n---\n(?P<body>.*)\Z", re.DOTALL)
 LINK = re.compile(r"(?<!!)\[(?P<label>[^\]]+)\]\((?P<target>[^)]+)\)")
 TAG = re.compile(r"^[a-z0-9]+(?:-[a-z0-9]+)*$")
-RECOMMENDED_CONCEPT_FIELDS = ("title", "description", "tags", "timestamp")
+ACTOR = re.compile(r"^(?:[a-z][a-z0-9-]*:[^\s:]+|[^/\s]+/[^/\s]+)$")
+ATX_H1 = re.compile(r"^ {0,3}#(?:[ \t]+(?P<text>.*))?[ \t]*$")
+TRAILING_ATX_CLOSE = re.compile(r"(?:^|\s)#+[ \t]*$")
+FENCE_LINE = re.compile(r"^ {0,3}(?P<marker>`{3,}|~{3,})")
+ISO_DATE = re.compile(r"^\d{4}-\d{2}-\d{2}$")
+RECOMMENDED_CONCEPT_FIELDS = ("title", "description", "tags", "generated", "sources")
```
Comment (dceoy) [inline src/aims/okf_hugo.py:44]: **[P2] Reject backticks in backtick-fence info strings**

This opener regex accepts arbitrary trailing text for both fence types. CommonMark forbids backtick characters in the info string of a backtick fence, so a body such as `# Computation` followed by an opener like `````python`invalid`` and a normal closer is counted as a valid sanctioned computation even though it is not a fenced code block under CommonMark. Distinguish backtick openers from tilde openers and reject a backtick-containing info string; add a regression case for this malformed opener.

**Review by dceoy — COMMENTED** [review]
The migration is well tested and CI is green, but two validation gaps still allow non-portable or non-CommonMark contracts to pass `--check`.

File: src/aims/okf_hugo.py (hunk 13 above)
Comment (dceoy) [reply src/aims/okf_hugo.py]: Fixed: sources with usage_count now require a document-level 'usage_window' regardless of any source-level window, which is validated only as a preserved extension (commit 8352caa).

File: src/aims/okf_hugo.py (hunk 14 above)
Comment (dceoy) [reply src/aims/okf_hugo.py:44]: Fixed: fence detection now treats a backtick-fenced opener whose info string contains a backtick as not a fence at all, per CommonMark, via a shared _fence_opener helper; tilde openers are unaffected (commit 8352caa).

**Review by dceoy — COMMENTED** [review]
Reviewed head `8352caa`. No actionable findings.

The OKF v0.2 migration, metadata validation/rendering behavior, curator provenance normalization, generated Hugo content, and regression coverage are consistent with #160. All prior review threads are resolved, and CI/CD run 303 passes.

**Review by dceoy — COMMENTED** [review]
The migration is broadly well tested, but two v0.2 provenance semantics are still incorrect: source-level `usage_window` overrides are rejected, and curator drafts backdate `generated.at` from the analysis reference date. Both can produce or reject materially incorrect metadata and should be fixed before merge.

File: src/aims/okf_hugo.py [hunk 15]
```diff
@@ -336,6 +346,429 @@ def validate_reserved(document: OkfDocument, src_root: Path) -> list[str]:
...
+                or isinstance(source["usage_count"], bool)
+                or source["usage_count"] < 0
+            ):
+                errors.append(
+                    _metadata_error(
+                        document,
+                        f"'{field}.usage_count' must be a non-negative integer",
+                    )
+                )
+            if "usage_window" not in document.metadata:
```
Comment (dceoy) [inline src/aims/okf_hugo.py]: **[P2] Allow a source-level usage window override**

OKF v0.2 §5.1 explicitly allows a single `sources` entry to carry its own `usage_window` to override the shared document-level window. This condition nevertheless requires `document.metadata["usage_window"]` whenever `usage_count` is present, so a conformant source such as `{resource: ..., usage_count: 1, usage_window: {from: ..., to: ...}}` is rejected by `--check`. Accept either the source-level window or the shared document-level window, and add a regression covering the source-only form.

Specification: https://github.com/GoogleCloudPlatform/knowledge-catalog/blob/main/okf/SPEC.md#51-provenance-sources

File: src/aims/okf_curation.py [hunk 16]
```diff
@@ -292,12 +301,13 @@ def concept_skeleton(cluster: ThemeCluster, as_of: str) -> str:
...
         "type: concept",
         f"tags: [{THEME_TAG}]",
-        f"timestamp: {as_of}T00:00:00Z",
-        "resource:",
-        f"  path: okf/concepts/theme-{slug}.md",
-        "  source: repository",
-        "status: proposed",
+        "generated:",
+        "  by: process:aims-okf-curator",
+        f"  at: {as_of}T00:00:00Z",
```
Comment (dceoy) [inline src/aims/okf_curation.py]: **[P2] Do not derive `generated.at` from the analysis `as_of` date**

`as_of` is the retirement/reference date (and defaults to the latest artifact date), not when this draft content was generated. Running the curator today with a historical `--as-of` value therefore emits a backdated `generated.at`, even though OKF v0.2 defines it as the current content's last meaningful change. Keep `as_of` for recurrence/retirement calculations and supply a separate generation timestamp (or an explicit placeholder that must be updated before committing).

Specification: https://github.com/GoogleCloudPlatform/knowledge-catalog/blob/main/okf/SPEC.md#52-trust-generated-and-verified

File: src/aims/okf_hugo.py (hunk 15 above)
Comment (dceoy) [reply src/aims/okf_hugo.py]: Accepted a source-level `usage_window` as an alternative to the document-level one in `_validate_sources()` (src/aims/okf_hugo.py:598), per OKF v0.2 §5.1; added a regression covering the source-only override (commit c856ff2).

File: src/aims/okf_curation.py (hunk 16 above)
Comment (dceoy) [reply src/aims/okf_curation.py]: `concept_skeleton()`/`render_proposal()` no longer derive `generated.at` from `as_of`; `render_proposal` now takes an optional `generated_at` defaulting to the current UTC time, with a regression test (commit c856ff2).

**Review by dceoy — COMMENTED** [review]
Reviewed the current head (`c856ff2`). No additional actionable findings.

The OKF v0.2 metadata validation, Hugo shadow mapping, curator provenance generation, and regression coverage are consistent with the v0.2 specification and the repository's stricter authoring policy. All prior review threads are resolved, and the CI/CD run for this head completed successfully.
